### ISSUE TITLE
Core: Dependency Graph: Graphviz failed to create an image file

### ISSUE THREAD

**Issue by AIRCAP** [FreeCAD/FreeCAD#17835]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

**Steps to reproduce:**
1. Unzip and open Fail.FCStd
[Fail.zip](https://github.com/user-attachments/files/17731459/Fail.zip)
2. Click on Tools->Dependency Graph

**Expected Outcome:**
Dependency Graph displayed

**Actual Outcome**
Popup: "Graphviz failure: Graphviz failed to create an image file.
![image](https://github.com/user-attachments/assets/cd0483d4-77f6-45d8-921b-0f3d4695d3a1)

**Workaround**
Exporting the dependency graph in graphviz format works. Theres only a few warnings:
Tools->Export Dependency Graph ---> test.gv
```bash
~$ dot -T png test.gv -o test.png
Warning: node '49', graph '' size too small for label
Warning: node '25', graph '' size too small for label
Warning: node '29', graph '' size too small for label
Warning: node '11', graph '' size too small for label
Warning: node '47', graph '' size too small for label
Warning: node '37', graph '' size too small for label
Warning: node '53', graph '' size too small for label
Warning: node '48', graph '' size too small for label
Warning: node '30', graph '' size too small for label
Warning: node '61', graph '' size too small for label
Warning: node '50', graph '' size too small for label
Warning: node '3', graph '' size too small for label
Warning: node '28', graph '' size too small for label
Warning: node '6', graph '' size too small for label
Warning: node '41', graph '' size too small for label
Warning: node '10', graph '' size too small for label
Warning: node '12', graph '' size too small for label
```
![image](https://github.com/user-attachments/assets/e406ac08-f766-45e3-b0f8-4557d3aa42fe)

### Full version info

```shell
Latest main:

OS: Ubuntu 20.04.6 LTS (XFCE/xfce/xcb)
Architecture: x86_64
Version: 1.1.0dev.39188 (Git)
Build type: Debug
Branch: main
Hash: 4bc8cd60e9a9c24ed007f0d3d341ca70498c798a
Python 3.8.10, Qt 5.12.8, Coin 4.0.0, Vtk 7.1.1, OCC 7.6.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/Qt default
Installed mods: 
  * Defeaturing 1.2.2
  * Silk 0.1.5
  * CurvedShapes 1.0.12
  * CfdOF 1.27.13
  * Curves 0.6.51
  * Plot 2024.10.23
  * AirPlaneDesign 0.4.1
  * lattice2 1.0.0
```

### Subproject(s) affected?

Core

### Anything else?

Other files work fine, its just this particular dependency graph that fails. If "Line1" is deleted, Graph display works.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by AIRCAP** [comment]
I ran into this when testing this new macro:
https://github.com/FreeCAD/FreeCAD-macros/pull/180

**Comment by maxwxyz** [comment]
Could this be related to the recently merged https://github.com/FreeCAD/FreeCAD/pull/16427 ? @hasecilu

**Comment by adrianinsaval** [comment]
does the appimage work?

**Comment by hasecilu** [comment]
Works on the weekly build 39176 from 18 hours ago, Arch Linux Wayland.

running `dot -Tsvg -Granksep=2 -Goutputorder=edgesfirst -Gsplines=ortho file.gv -o file.svg` works?

If you say it's only that file it's not related to recent changes, I gues you have self compiled version, could you try the AppImage?


Your file

![image](https://github.com/user-attachments/assets/6aac26a9-4e79-4845-aeb7-b7bf2c1afa2e)

Another file

![image](https://github.com/user-attachments/assets/95de408e-dc77-45a6-a5cd-6c34c3acc8fa)

**Comment by AIRCAP** [comment]
I managed to get a test run of the ApImage and I tested RC4 and the latest weekly. Error doesn't happen there, only in self compiled main. Any idea what could cause that? Is FreeCAD using some weird graphviz python library that could be different from the system wide dot?

**Comment by adrianinsaval** [comment]
> Is FreeCAD using some weird graphviz python library

noup, but the appimage includes it's own version of graphviz. Are both dot and unflatten available in your path?

**Comment by AIRCAP** [comment]
> > Is FreeCAD using some weird graphviz python library
> 
> noup, but the appimage includes it's own version of graphviz. Are both dot and unflatten available in your path?

Yes. As marked before, it the graph generation works fine with virtually all other FCStd files I have - including this one if that BSpline called "Line" is removed.

```shell
$ which unflatten
/usr/bin/unflatten
$ which dot
/usr/bin/dot
$ dot -V
dot - graphviz version 2.43.0 (0)
```

unflatten doesn't have a `-V` flag for version, but since its bundled I assume its the same

**Comment by hasecilu** [comment]
I also have a self-compiled Qt6 version and also worked there, not sure what have could be wrong, if it's no causing troubles on other files and can't be reproduce will be hard to follow.

**Comment by dzid26** [comment]
Dependency Graph never worked for me in various versions of Freecad.
I am not sure if height/width is a problem, since the dot works with it anyway

```
digraph "" {
graph [
compound=true];
subgraph clusterBody {
graph [
bgcolor="#69e2c550", label=Body, style="rounded,filled"];
subgraph clusterOrigin {
graph [
bgcolor="#15ba7050", label=Origin, style="rounded,filled"];
}
subgraph clusterOrigin {
graph [
bgcolor=none, label=Origin];
1[label="YZ_Plane&#92;n(YZ-plane)", shape=Mrecord, style=filled];
2[label="XY_Plane&#92;n(XY-plane)", shape=Mrecord, style=filled];
3[fixedsize=true, height=0, style=invis, width=0];
4[label="X_Axis&#92;n(X-axis)", shape=Mrecord, style=filled];
5[label="Y_Axis&#92;n(Y-axis)", shape=Mrecord, style=filled];
6[label="Z_Axis&#92;n(Z-axis)", shape=Mrecord, style=filled];
7[label="XZ_Plane&#92;n(XZ-plane)", shape=Mrecord, style=filled];
8[label=Origin001, shape=Mrecord, style=filled];
}
0[fixedsize=true, height=0, style=invis, width=0];
9[label=Sketch, shape=Mrecord, style=filled];
10[label=Pad, shape=Mrecord, style=filled];
9 -> 2;
10 -> 9[label=" 2x"];
}
}

```


`dot -Tsvg -Granksep=2 -Goutputorder=edgesfirst -Gsplines=ortho cube.gv -o cube.svg`
`Warning: node '3', graph '' size too small for label`
`Warning: node '0', graph '' size too small for label`
`Warning: Orthogonal edges do not currently handle edge labels. Try using xlabels.`
![cube](https://github.com/user-attachments/assets/031889f4-85d4-4e2c-aeb7-fd478a749769)


```
 dot -v
dot - graphviz version 2.50.0 (20211204.2007)
libdir = "C:\Program Files\Graphviz\bin"
Activated plugin library: gvplugin_dot_layout.dll
Using layout: dot:dot_layout
Activated plugin library: gvplugin_core.dll
Using render: dot:core
Using device: dot:dot:core
The plugin configuration file:
        C:\Program Files\Graphviz\bin\config6
                was successfully loaded.
    render      :  cairo dot dot_json fig gdiplus json json0 map mp pic pov ps svg tk vml xdot xdot_json
    layout      :  circo dot fdp neato nop nop1 nop2 osage patchwork sfdp twopi
    textlayout  :  textlayout
    device      :  bmp canon cmap cmapx cmapx_np dot dot_json emf emfplus eps fig gif gv imap imap_np ismap jpe jpeg jpg json json0 metafile mp pdf pic plain plain-ext png pov ps ps2 svg tif tiff tk vml xdot xdot1.2 xdot1.4 xdot_json
    loadimage   :  (lib) bmp eps gif jpe jpeg jpg png ps svg
```

**Comment by leoheck** [comment]
I am still seeing this issue.

The generated file has fixedsize `with =0`  maybe the decimal places were cut from the out put.

<img width="854" height="647" alt="Image" src="https://github.com/user-attachments/assets/e258d1f3-0964-448e-b2a3-0feef91da14a" />


Then this simple command would fail.

```
/tmp/fc 
✦ ➜ dot -Tsvg lala.gv -o lala.svg
Warning: node '2', graph '' size too small for label
Warning: node '5', graph '' size too small for label
```

A quick fix to update the hardcoded font size fixes the file.
<img width="854" height="647" alt="Image" src="https://github.com/user-attachments/assets/7e33ddb6-10dc-4792-a5a1-319d164c5f0a" />

<img width="579" height="160" alt="Image" src="https://github.com/user-attachments/assets/7c3a29b2-6936-4881-9671-fcb5ffebe558" />

<img width="632" height="578" alt="Image" src="https://github.com/user-attachments/assets/b7504cfa-98ac-41de-a84c-41bedfbdd829" />

```
OS: Ubuntu 25.04 (ubuntu:GNOME/ubuntu-xorg/xcb)
Architecture: x86_64
Version: 1.1.0dev.14555 (Git shallow)
Build date: 2025/09/12 03:41:38
Build type: Release
Branch: main
Hash: 952bff5b63a62b55a7ea1623edfb39e90dee2d4c
Python 3.11.13, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.3.0, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: English/United States (en_US)
Navigation Style/Orbit Style/Rotation Mode: Touchpad/Trackball/Drag at cursor
Stylesheet/Theme/QtStyle: OpenDark.qss/OpenDark/
Logical DPI/Physical DPI/Pixel Ratio: 96/94.0767/1
 ```

**Comment by andershol** [comment]
The "size too small for label" warnings appear to be do to the "fixedsize=true, height=0, style=invis, width=0", as the nodes mentioned are those with those attributes. But [the code that produces those attributes](https://github.com/FreeCAD/FreeCAD/blob/main/src/App/Graphviz.cpp#L341) writes them as hard-coded strings (ie. there is no rounding going on) and is from 2017 (ignoring code reformatting), ie. it is not new, so the warnings are probably unrelated to the "Graphviz failed to create an image file" and can be ignored.

When selecting "Dependency Graph" the code first produces the same gv code as is produced when you select "Export Dependency Graph", this is then preprocessed with the command "unflatten -c2 -l2" (unless the "unflatten" parameter is set to false) and then converted to a svg with "dot -Granksep=2 -Goutputorder=edgesfirst -Gsplines=ortho -Tsvg". Doing this with e.g. graphviz 2.38 (which was coincidentally the version I had, and appear close to the 2.43.0 version mentioned above) you get a crash in the dot command and no output is produced (the command takes a few seconds to run and on windows the "WerFault.exe" program starts to write a dmp file i the "%LOCALAPPDATA%\CrashDumps" folder). So the missing image is due to graphviz crashing.

There didn't immediately seem to be something wrong with the gv file, so I tried minimizing the gv file to a minimal file that produced the crash. Depending on which order I removed the elements I produced different minimized file, but one can be (and the crash produced by this minimal might of course not be the same crash as from the full file):
```
digraph  {
	graph[compound=true];
	subgraph clusterOrigin002 { 32; 33[shape=Mrecord]; }
	subgraph clusterPart1 { 15[fixedsize=true, height=0, width=0]; 47[label=Placement]; }
	12 -> 15[lhead=clusterPart1];
	54 -> 15[lhead=clusterPart1];
	57 -> 15[lhead=clusterPart1];
	32 -> 33;
}
```
This do not appear inherently wrong to me, and using a current graphviz (14.1.4) produces a svg just fine. And making FreeCAD use this newer version of graphviz, fixes the problem on my machine. So I believe that the crash is due to a bug in earlier versions from graphviz and that this bug will be difficult for FreeCAD to work around. The correct solution is to update your graphviz and make sure that it is being used by FreeCAD by checking the "BaseApp/Preferences/Paths/Graphviz" parameter (on linux the code falls back to looking in "/usr/bin").

Maybe the "Graphviz failed to create an image file" dialog should just have text added to encorange the user to upgrade their graphviz, maybe it could clear the "BaseApp/Preferences/Paths/Graphviz" parameter so the user is prompted for a new location next time, or the value could be shown or "dot --version" could be run and the version shown.

Edit: I realized that my sample contained the "fixedsize=true" attributes, so that wasn't a good way to argue that is was unrelated (the argument still holds because of the "is from 2017" stuff). So I made a small script to minimize the gv in a randomized way and most of them contained "shape=Mrecord" while few contained "fixedsize=true", eg.:
```
digraph {
	graph[compound=true];
	subgraph cluster1 { 32; 33; 35[shape=Mrecord]; }
	subgraph cluster2 { 15; }
	52 -> 12 -> 15;
	57 -> 15[lhead=cluster2];
	57 -> 56;
	32 -> 33 -> 35;
}
```

**Comment by Sil68** [comment]
> The correct solution is to update your graphviz and make sure that it is being used by FreeCAD by checking the "BaseApp/Preferences/Paths/Graphviz" parameter (on linux the code falls back to looking in "/usr/bin").
> 

This parameter was not set at all. -- After adding it all works as intended!

Thanks mate! :)

**Comment by FlachyJoe** [comment]
try
```type unflatten```
if result is `not found`
then install the package _graphviz-tools_ (debian trixie+) or similar  from your package manager.

**Comment by Sil68** [comment]
> [@Sil68](https://github.com/Sil68) can you export the Graphviz file, and share it here? Does it export it?

I managed to successfully export the graph

<img width="640" height="483" alt="Image" src="https://github.com/user-attachments/assets/c08047c3-7c30-4f45-8da3-a5e633624658" />
